src/uri.jl

```diff
@@ -153,46 +153,24 @@ function Base.isvalid(uri::URI)
 end
 
 # RFC3986 Unreserved Characters (and '~' Unsafe per RFC1738).
-@inline shouldencode(c) = !((c >= UInt8('A') && c <= UInt8('Z'))
-                         || (c >= UInt8('a') && c <= UInt8('z'))
-                         || (c >= UInt8('0') && c <= UInt8('9'))
-                         || c == UInt8('-')
-                         || c == UInt8('.')
-                         || c == UInt8('_'))
-hexstring(x) = string('%', uppercase(hex(x,2)))
+@inline isunsafe(c::Char) = !(c in "-._") && (!isascii(c) || !isalnum(c))
+
+utf8_chars(str::AbstractString) = (Char(c) for c in Vector{UInt8}(str))
 
 "percent-encode a string, dict, or pair for a uri"
 function escape end
 
-escape(v::Number) = string(v)
-escape(v::Nullable{T}) where {T} = Base.isnull(v) ? "" : string(get(v))
-function escape(str::AbstractString, f=shouldencode)
-    out = IOBuffer()
-    for c in Vector{UInt8}(str)
-        write(out, f(c) ? hexstring(Int(c)) : c)
-    end
-    return String(take!(out))
-end
-escape(bytes::Vector{UInt8}) = bytes
+escape(c::Char) = string('%', uppercase(hex(c,2)))
+escape(str::AbstractString, check::Function=isunsafe) = 
+    join(check(c) ? escape(c) : c for c in utf8_chars(str))
 
-escape(io, k, v) = write(io, escape(k), "=", escape(v))
-function escape(io, k, A::Vector{String})
-    len = length(A)
-    for (i, v) in enumerate(A)
-        write(io, escape(k), "=", escape(v))
-        i == len || write(io, "&")
-    end
-end
+escape(bytes::Vector{UInt8}) = bytes
```

Comment:
(Note: the raw bytes were not escaped before this PR either.)
I assumed that it was intended as a way to pass in bytes that are not to be escaped. e.g. if you have some pre-escaped data. Other than that, I can't see a use-case for passing raw bytes as part of a URL or query string.